### PR TITLE
[niconico] Support encrypted official videos

### PR DESCRIPTION
**Pull request by tsukumijima**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [x] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

ytdl-org#23824 .

I made this pull request before, but I was giving up because it didn't merge at all.
Recently, when I was looking at the issues of youtube-dl, I found out that a fork called youtube-dlc was made, and I made a pull request.
(Since it uses Google Translate, the sentence may be unnatural)

### CODE REVIEW

**Comment by blackjack4494** [comment]
Will have a look at it ~tomorrow. The heartbeat can probably be implemented directly into the extractor or as another function.

**Comment by tsukumijima** [comment]
I think that the implementation of heartbeat used the code of https://github.com/panhartstuff/youtube-dl/commit/ff1968568a32cbb6d64440075124faefdff287f6.
It may be possible to implement better. (I'm not familiar with Python...)

I think that ytdl-org#14582 etc. can be helpful.

**Comment by blackjack4494** [comment]
Some collection
https://github.com/ytdl-org/youtube-dl/pull/23824
https://github.com/ytdl-org/youtube-dl/issues/14582
https://github.com/ytdl-org/youtube-dl/pull/16495


Hm. It seems trickier than I thought. In theory this should work and not affect any other extractors since you check for `heartbeat_url` in `info_dict`. Maybe it's just me but code structure doesn't feel natural. There are unfortunately multiple ways to handle this.  
- Like it is in this PR
- This rather (imo) poor attempt https://github.com/ytdl-org/youtube-dl/issues/14582#issuecomment-360767112
  - but still got potential to be included into the extractor itself
- extend the http downloader https://github.com/ytdl-org/youtube-dl/issues/14582#issuecomment-362788161
  - it's similar to the `youtube_live_chat` `downloader` `YoutubeLiveChatReplayFD` which seems like a good approach

Short term fix would be to actually merge this PR. However it's kind of bad practice to change the core like this.

**Comment by tsukumijima** [comment]
As you said, I agree that it is best not to change the core as much as possible.

However, niconico's videos use both HTTP and HLS, of which the official encrypted videos use only HLS.
Also, as https://github.com/ytdl-org/youtube-dl/pull/16495/files#r189578412, this implementation cannot be used with other stream types such as HLS.
Therefore, implementations that extend HttpFD, such as the third YoutubeLiveChatReplayFD, will at least need to make further changes to the code in the comments.

If you don't want to change the core, it would be nice to improve the implementation that does the heartbeat in the second Extractor, but what about?

**Comment by bbepis** [comment]
A lot of niconico improvements (this and others) were added in another fork: https://github.com/animelover1984/youtube-dl

I've tried the other implementations, and I couldn't find a better solution than this one.

I will say though; adding support for this website is a massive pain in the ass. A combination of them trying to lock downloaders out and bad coding of the website in general require a lot of changes to the core of youtube-dl to get things working at all.
For example, adding the ability to download livestreams require websockets, which itself requires a pip package and python 3.7+.

This is probably the only change for this site that still allows youtube-dl to be run on python 2

**Comment by blackjack4494** [comment]
> A lot of niconico improvements (this and others) were added in another fork: https://github.com/animelover1984/youtube-dl
> 
> I've tried the other implementations, and I couldn't find a better solution than this one.
> 
> I will say though; adding support for this website is a massive pain in the ass. A combination of them trying to lock downloaders out and bad coding of the website in general require a lot of changes to the core of youtube-dl to get things working at all.
> For example, adding the ability to download livestreams require websockets, which itself requires a pip package and python 3.7+.
> 
> This is probably the only change for this site that still allows youtube-dl to be run on python 2

Big Thanks will have a look at it. There are websocket packages for python 2.7 tho. But I generally advise against python2 anyway. EOL will be January 1st 2020 so pretty soon.  
I honestly don't know why this site tries to use such weird mechanisms. They can be reversed. DRM would likely be a better option but hard to implement in large scale and can be defeated at some point.

While looking at your code. Is websockets really needed for heartbeat? Those alternative options don't use websockets. Is it just out of convenience or actually required?

**Comment by bbepis** [comment]
> While looking at your code. Is websockets really needed for heartbeat? Those alternative options don't use websockets. Is it just out of convenience or actually required?

No, I was just a bit unclear. What I meant was that improvements related to this site that weren't this heartbeat issue involve websockets.

**Comment by tsukumijima** [comment]
Supplement from me (I'm not good at English and I read using Google Translate, so I may have misunderstood the meaning).

First, you don't need WebSockets to download AES-128 encrypted videos. Also, sending heartbeats does not require WebSockets. The heartbeat is done with a regular HTTP request.

If you want to add a "niconico live" extractor, like the code at https://github.com/animelover1984/youtube-dl, you'll need a WebSocket. "Nico nico live" sends information via WebSocket and seems to be needed when downloading.
Currently, a tool called "Nico Live New Distribution Recording Tool (Tentative)" (https://github.com/guest-nico/nicoNewStreamRecorderKakkoKari) is often used to download "niconico live". For the specifications, https://github.com/guest-nico/nicoNewStreamRecorderKakkoKari/blob/master/siyou.md will be helpful (although all are in Japanese ...).
"Niconico video (ニコニコ動画)" and "niconico live (ニコニコ生放送)" are operated by the same company under the same "niconico (ニコニコ)" brand, but they use completely different systems.

> Big Thanks will have a look at it. There are websocket packages for python 2.7 tho. But I generally advise against python2 anyway. EOL will be January 1st 2020 so pretty soon.

I don't understand why youtube-dl still supports Python2 beyond EOL ... Supporting both Python2 and Python3 makes development, especially support for complex sites like "niconico" more difficult. I will. (Python3 supports up to Python 3.2 released in 2011)
I think at least Python2 support should be discontinued.

### About encrypted videos
> I honestly don't know why this site tries to use such weird mechanisms. They can be reversed. DRM would likely be a better option but hard to implement in large scale and can be defeated at some point.

First, about encrypted videos.
Applies only to some official videos such as anime.
For the mechanism, see https://qiita.com/tor4kichi/items/91550a71119f3878bfba#comment-346de200d84c79876831 (Japanese).
The mechanism of encryption itself is common, but the URL of the decryption key can only be accessed once. If you access it more than once, you will not be able to decrypt the segment because it will return a fake key that cannot be decrypted.
In addition, you need to ping https://nvapi.nicovideo.jp/v1/2ab0cbaa/watch?t={tracking_id} before downloading the playlist. Otherwise, a fake key that cannot be decrypted will be returned, even if the decryption key has never been accessed.
Previously, no measures were taken against downloads, but it has been implemented since March 2019. However, even official videos may not be encrypted.
I presume that "niconico video" intentionally does not use DRM. This is because there is a license fee to use DRM (the operating company of "niconico" has been in the red until recent years), and browser support is limited.
In the fix / niconico-c branch, processing has been added to handle these.

### About heartbeat
Next, about heartbeat.
The mechanism is https://qiita.com/tor4kichi/items/91550a71119f3878bfba#%E5%8B%95%E7%94%BB%E5%86%8D%E7%94%9F%E3%81%AE%E3%83 % 8F% E3% 83% BC% E3% 83% 88% E3% 83% 93% E3% 83% BC% E3% 83% 88% E3% 82% 92% E8% A1% 8C% E3% 81% 86 Please see (Japanese).
Heartbeat is required for all videos that use the new Nico Nico Douga distribution system (DMC: Dwango Media Cluster), with or without encryption.
Old videos may use the old distribution system, but it seems that they will gradually move to the new distribution system.
The video viewing session itself can be obtained by analyzing the json data in the source of the viewing page. However, if you do not send a heartbeat, the viewing session will be closed after a certain amount of time.
Short videos can be downloaded while the viewing session is active, but long videos will fail to download in the middle because the viewing session will be closed during the download.
Unencrypted videos use HTTP, so youtube-dl's resume feature allows you to resume the download from the middle of the download. This means that you can run youtube-dl a few times to complete the download. However, it is very painful to keep running youtube-dl over and over again.
Encrypted videos use HLS and cannot be resumed because the decryption key is different for each viewing session.
In other words, in order to support encrypted videos, in addition to complying with the specifications described in "About encrypted videos", send heartbeats regularly during download to maintain the viewing session Must be.
If you implement the heartbeat function, in addition to being able to support encrypted videos, you will be able to download unencrypted videos with just one run of youtube-dl without resuming.

I don't understand why "niconico video" uses such a strange mechanism. Is it a countermeasure against downloads, or is it to reduce the load on traffic and servers ...

**Comment by blackjack4494** [comment]
Haven't forgotten this PR.

**Comment by rivke41levp656** [comment]
This is great. The only problem I have is  `uploader` metadata is not fetched or some other problem —it becomes `NA` in output template. With the same output template,  aforementioned fork (https://github.com/animelover1984/youtube-dl) produces correct output.

**Comment by tsukumijima** [comment]
Long time no see!
We are pleased that each other's repositories are back.
Currently it is being developed with "blackjack4494/yt-dlc", what will you do in the future?
I was temporarily developing with "tsukumijima/yt-dl", but since the original repository was restored, "tsukumijima/yt-dl" was archived.
